src/transformers/modeling_blenderbot.py

```diff
@@ -0,0 +1,228 @@
+            )
+        assert isinstance(encoder_outputs, BaseModelOutput)
+        if use_cache:
+            decoder_padding_mask, casual_mask = None, None
+        else:
+            decoder_input_ids, decoder_padding_mask, casual_mask = _prepare_bart_decoder_inputs(
+                self.config,
+                input_ids,
+                decoder_input_ids=decoder_input_ids,
+                causal_mask_dtype=self.shared.weight.dtype,
+                decoder_padding_mask=decoder_attention_mask,
+            )
+        assert decoder_input_ids is not None
+        decoder_outputs = self.decoder.forward(
+            decoder_input_ids,
+            encoder_outputs[0],
+            attention_mask,
+            decoder_padding_mask,
+            decoder_causal_mask=casual_mask,
+            decoder_past_key_values=decoder_cached_states,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            use_cache=use_cache,
+        )
+
+        #scores = F.linear(decoder_outputs[0], self.shared.weight)
+        scores = self.output(decoder_outputs[0])
+        # outputs = (scores,) + outputs[1:]
+        loss = None
+        if labels is not None:
+            loss_fc = nn.CrossEntropyLoss()
+            loss = loss_fc(scores[0].view(-1, self.config.vocab_size), labels.view(-1))
+
+        return Seq2SeqLMOutput(
+            loss=loss,
+            logits=scores,
+            decoder_past_key_values=decoder_outputs.past_key_values,
+            decoder_hidden_states=decoder_outputs.hidden_states,
+            decoder_attentions=decoder_outputs.attentions,
+            encoder_last_hidden_state=encoder_outputs.last_hidden_state,
+            encoder_hidden_states=encoder_outputs.hidden_states,
+            encoder_attentions=encoder_outputs.attentions,
+        )
+
+    def prepare_inputs_for_generation(self, decoder_input_ids, past, attention_mask, use_cache, **kwargs):
+        # exactly as in BartConditionalGeneration
+        assert past is not None, "past has to be defined for encoder_outputs"
+        # first step, decoder_cached_states are empty
+        encoder_outputs, decoder_cached_states = past
+        return {
+            "input_ids": None,
+            "encoder_outputs": encoder_outputs,
+            "decoder_cached_states": decoder_cached_states,
+            "decoder_input_ids": decoder_input_ids,
+            "attention_mask": attention_mask,
+            "use_cache": use_cache,
+        }
+
+    def get_input_embeddings(self):
+        return self.shared
+
+    def set_input_embeddings(self, value):
+        self.shared = value
+        self.encoder.embed_tokens = self.shared
+        self.decoder.embed_tokens = self.shared
+
+    def get_output_embeddings(self) -> nn.Linear:
+        vocab_size, embed_dim = self.shared.weight.shape
+        lin_layer = nn.Linear(vocab_size, embed_dim, bias=False)
```

Comment:
Same as in Bart -> Don't think it's a good practice to generate the word embedding on the fly...we should reuse the code as defined in `PretrainedModels` that "ties" the output weights to the input weights

Reply:
I will fix it in bart separately.